src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpDocumentExcerptService.cs

```diff
@@ -0,0 +1,148 @@
+        {
+            if (mappingService is null)
+            {
+                throw new ArgumentNullException(nameof(mappingService));
+            }
+
+            if (documentSnapshot is null)
+            {
+                throw new ArgumentNullException(nameof(documentSnapshot));
+            }
+
+            _mappingService = mappingService;
+
+            _documentSnapshot = documentSnapshot;
+        }
+
+        internal override async Task<ExcerptResultInternal?> TryGetExcerptInternalAsync(
+            Document document,
+            TextSpan span,
+            ExcerptModeInternal mode,
+            CancellationToken cancellationToken)
+        {
+            var mapped = await _mappingService.MapSpansAsync(document, new[] { span }, cancellationToken).ConfigureAwait(false);
+            if (mapped.Length == 0 || mapped[0].Equals(default(RazorMappedSpanResult)))
+            {
+                return null;
+            }
+
+            var razorDocumentText = _documentSnapshot.Snapshot.AsText();
+            var razorDocumentSpan = razorDocumentText.Lines.GetTextSpan(mapped[0].LinePositionSpan);
+
+            var generatedDocument = document;
+
+            // First compute the range of text we want to we to display relative to the razor document.
+            var excerptSpan = ChooseExcerptSpan(razorDocumentText, razorDocumentSpan, mode);
+
+            // Then we'll classify the spans based on the razor document, since that's the coordinate
+            // space that our output mappings use.
+            var classifiedSpans = await ClassifyPreviewAsync(
+                razorDocumentSpan,
+                excerptSpan,
+                span,
+                generatedDocument,
+                cancellationToken).ConfigureAwait(false);
+
+            var excerptText = GetTranslatedExcerptText(razorDocumentText, ref razorDocumentSpan, ref excerptSpan, classifiedSpans);
+
+            return new ExcerptResultInternal(excerptText, razorDocumentSpan, classifiedSpans.ToImmutable(), document, span);
+        }
+
+        private async Task<ImmutableArray<ClassifiedSpan>.Builder> ClassifyPreviewAsync(
+            TextSpan razorSpan,
+            TextSpan excerptSpan,
+            TextSpan generatedSpan,
+            Document generatedDocument,
+            CancellationToken cancellationToken)
+        {
+            var builder = ImmutableArray.CreateBuilder<ClassifiedSpan>();
+
+            var remainingSpan = excerptSpan;
+
+            // We should be able to process this whole span as C#, so classify it.
+            //
+            // However, we'll have to translate it to the the generated document's coordinates to do that.
+            var offsetRazorToGenerated = generatedSpan.Start - razorSpan.Start;
+            var offsetExcerpt = new TextSpan(excerptSpan.Start + offsetRazorToGenerated, excerptSpan.Length);
+
+            var classifiedSecondarySpans = await Classifier.GetClassifiedSpansAsync(
```

Comment:
I feel like this area needs to be called out a bit, specifically (https://github.com/dotnet/aspnetcore-tooling/pull/2041/commits/105efe652cf1bcc3c805a9e425528c8f0d5d343a).

Essentially, we're putting the whole razor excerpt span through the C# classifier (including HTML). From my testing, this doesn't really seem to hurt (HTML just gets colored gray, and shows up as gray in the preview).

The reason for doing this, is without it, only the method name for instance gets colorized on the reference item, and we treat the rest of the line as `Text` (no colorization for comments, types, etc.). This issue becomes especially pronounced for previews (hovering over the item), in which case without this change, everything shows up gray.

Reply:
Ahh thanks for calling this out. Is the reason why we don't pass in just the C# just because we haven't written the logic to extract the C# pieces from a given portion of the Razor document?

Reply:
Yep! If/when we can extract the C# `TextSpan`(s) from a Razor `TextSpan`, this can be updated to only request for classification on the C# (sub-)spans.

Reply:
Gotcha! That sounds fine for now if you ask me.